### PR TITLE
Add receiver downscale endpoint

### PR DESCRIPTION
**Pull request by yuchen-db**
The endpoint expose number of TSDBs, rollout operator will patch the sts if n_tsdb=0
related PRs:
https://github.com/databricks/universe/pull/753653
https://github.com/databricks/rollout-operator/pull/7

### CODE REVIEW

File: pkg/server/http/http.go [hunk 1]
```diff
@@ -17,6 +15,11 @@ import (
 	toolkit_web "github.com/prometheus/exporter-toolkit/web"
 	"golang.org/x/net/http2"
 	"golang.org/x/net/http2/h2c"
+	"net/http"
```
Comment (jnyi) [inline pkg/server/http/http.go]: please fix the import convention for go: https://google.github.io/styleguide/go/best-practices.html#import-ordering

File: cmd/thanos/receive.go [hunk 2]
```diff
@@ -311,6 +311,8 @@ func runReceive(
 			httpserver.WithGracePeriod(time.Duration(*conf.httpGracePeriod)),
 			httpserver.WithTLSConfig(*conf.httpTLSConfig),
 		)
+		var lastDownscalePrepareTimestamp *int64 = nil
```
Comment (jnyi) [inline cmd/thanos/receive.go]: this implementation assumes the pod won't restart, to preserve the state, we might need to persist it on local PVC too, any thoughts?

File: pkg/receive/multitsdb.go [hunk 3]
```diff
@@ -106,6 +106,14 @@ func (t *MultiTSDB) SkipMatchExternalLabels() {
 	t.skipMatchExternalLabels = true
 }
 
+func (t *MultiTSDB) GetTenants() map[string]*tenant {
+	return t.tenants
```
Comment (hczhu-db) [inline pkg/receive/multitsdb.go]: This map is protected by `mtx`. A call site of `GetTenants()` can hold a pointer to the map and do reads/writes while the map is being updated by other goroutines. That'd be a data race.
It's safer to return a deep copy of the map if copying is not too expensive.

File: pkg/server/http/http.go [hunk 4]
```diff
@@ -68,6 +71,29 @@ func New(logger log.Logger, reg *prometheus.Registry, comp component.Component,
 	}
 }
 
+// RegisterDownscale registers HTTP handler compatible with Grafana rollout operator.
+// See https://github.com/databricks/rollout-operator?tab=readme-ov-file#delayed-scaledown.
+func RegisterDownscale[K comparable, V any](s *Server, m map[K]V, mtx *sync.RWMutex, t *int64) {
+	s.mux.Handle("/-/downscale", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		mtx.RLock()
+		n := len(m)
```
Comment (hczhu-db) [inline pkg/server/http/http.go]: Don't pass in a map, if only the size of map is needed. Pass in the size of the map.

File: cmd/thanos/receive.go (hunk 2 above)
Comment (yuchen-db) [reply cmd/thanos/receive.go]: yes but it's safe to lose this state. rollout operator will only do downscale if the returned timestamp is more than 30 seconds ago. If we lose this state, it returns the current timestamp and downscale will not accidentally happen. In the worst case, a scheduled downscale will be delayed for 30 seconds.

File: pkg/server/http/http.go (hunk 4 above)
Comment (yuchen-db) [reply pkg/server/http/http.go]: the size of the map changes dynamically during handler runtime and [Golang passes map by reference ](https://stackoverflow.com/questions/40680981/are-maps-passed-by-value-or-by-reference-in-go)

File: pkg/server/http/http.go (hunk 1 above)
Comment (yuchen-db) [reply pkg/server/http/http.go]: refactored

File: pkg/server/http/http.go (hunk 4 above)
Comment (yuchen-db) [reply pkg/server/http/http.go]: refactored code

File: pkg/receive/multitsdb.go (hunk 3 above)
Comment (yuchen-db) [reply pkg/receive/multitsdb.go]: refactored code

**Review by hczhu-db — APPROVED** [review]

File: pkg/receive/multitsdb.go [hunk 5]
```diff
@@ -122,6 +122,12 @@ func NewMultiTSDB(
 	return mt
 }
 
+func (t *MultiTSDB) GetTenantsLen() int {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return len(t.tenants)
```
Comment (jnyi) [inline pkg/receive/multitsdb.go]: shall we log how many tenants left and what are they?

File: cmd/thanos/receive.go [hunk 6]
```diff
@@ -7,8 +7,10 @@ import (
 	"context"
 	"fmt"
 	"net"
+  "net/http"
```
Comment (jnyi) [inline cmd/thanos/receive.go]: indent

File: pkg/receive/multitsdb.go (hunk 5 above)
Comment (jnyi) [reply pkg/receive/multitsdb.go]: nit: add unit test?

**Review by jnyi — APPROVED** [review]
lgtm, would be nice if you can add unit tests, especially after prune.

**Pull request merged**